<div style="background-color: white; width: 960px; margin:0 auto; text-align: left; border: 1px;">
  <h1>FAQs and additional information</h1>
  <hr>
  <br>
  <h2>FAQs</h2>
  <hr>
  <h3>How were sample metadata parsed to generate tissue- and disease-specific expression data?</h3>
  <p>Sample metadata was parsed using a tissue-specific regular expressions dictionary manually curated and validated for sanity against a randomized list of sample tissue metadata entries. This dictionary and accompanying pre-processing scripts can be downloaded from the <a href="https://github.com/millerh1/correlationAnalyzeR" target="_blank"><em>Correlation AnalyzeR</em> github repository</a>.</p>
  <br>
  <h3>How was the expression data processed into correlations?</h3>
  <p>The expression data was pre-processed by <a href="https://amp.pharm.mssm.edu/archs4/" target="_blank">ARCHS4</a> using the pipeline described in their <a href="https://www.nature.com/articles/s41467-018-03751-6" target="_blank">2018 publication</a>. They generously provide free access to their expression data matrices and many other useful resources through their <a href="https://amp.pharm.mssm.edu/archs4/download.html" target="_blank">downloads page</a>. From ARCHS4, human and mouse expression data were downloaded separately. Using sample metadata, samples from single-cell experiments were excluded. Additional keywords were used to parse normal and cancer samples, samples of different tissue types, and samples from different species. Normalization of count data proceeded using the procedure <a href="https://amp.pharm.mssm.edu/archs4/help.html" target="_blank">outlined by the ARCHS4 authors</a>: counts were log2(x+1) transformed, quantile normalized, and batch corrected by GEO study accession. Correlations were calculated using the <a href="https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/" target="_blank">WGCNA</a> <em>corr</em> function. The scripts used for pre-processing are also available in the <a href="https://github.com/millerh1/correlationAnalyzeR" target="_blank"><em>Correlation AnalyzeR</em> github repository</a>.</p>
  <br>
  <h3>How is GSEA implemented on gene correlations?</h3>
  <p>Geneset Enrichment Analysis (GSEA) is a versatile technique that calculates pathway enrichment for any genome-wide ranked gene list (specifically, the <a href="https://www.biorxiv.org/content/10.1101/060012v1" target="_blank">fGSEA method</a> is used here). An overview of this approach is provided in an excellent <a href="https://davetang.org/muse/2018/01/10/using-fast-preranked-gene-set-enrichment-analysis-fgsea-package/" target="_blank">blog post</a> by Dave Tang. Usually, gene lists are ranked by some output of differential gene expression analysis -- but any valid ranking metric can be used to sort a gene list prior to GSEA. In this case, a gene is chosen (e.g. BRCA1) which contains ~26k correlation values for every other gene in the genome. By ranking these correlated genes with their correlation values, a valid pre-ranked gene list is produced for GSEA. The top enriched pathways produced by fGSEA analysis can be considered co-expressed/correlated with the gene.</p>
  <br>
  <h3>How is significance testing implemented in paired-mode?</h3>
  <p>Paired-mode involves a user choosing one primary gene and a list of secondary genes to determine if the secondary gene list is correlated with the primary gene. The correlation of these genes is compared to random chance by using a permutation approach. In simple terms, the analysis asks "is the degree of correlation for the selected gene list greater than if someone were to randomly select the same number of genes?"<br>For each permutation, a list of random genes with the same length as the secondary gene list is chosen. Then, a two-sided t test is performed to determine whether the absolute correlation values of the randomly-chosen genes are significantly different from those of the secondary gene list. By performing 2000 permutations, a distribution of t test p values is generated. If this distribution forms a peak below p=0.05, then it is determined that the secondary gene list is significantly correlated with the primary gene compared to random chance. Absolute correlations were used so that strong correlations would be treated equally regardless of sign.</p>
  <br>
  <hr>
  <h2>Additional information</h2>
  <p>For website support and additional assistance please contact <a href="mailto:millerh1@livemail.uthscsa.edu">Henry Miller</a></p>
  <p>For bug reports please open an issue and someone will address it shortly<a href="https://github.com/millerh1/correlationAnalyzeR-ShinyApp/issues" target="_blank"> <span class="glyphicon glyphicon-share-alt" aria-hidden="true"></span> <img src="GitHub-Mark.png" height="20"></a></p>
  <br>
  <h4>The <em>correlationAnalyzeR</em> R package is available for download:</h4>
  <code>## install.packages("devtools")<br>devtools::install_github("millerh1/correlationAnalyzeR")</code>
</div>

